br/pkg/lightning/metric/metric.go

```diff
@@ -51,208 +53,255 @@ const (
+			}, []string{"name"}),
+
+		RowReadSecondsHistogram: factory.NewHistogram(
+			prometheus.HistogramOpts{
+				Namespace: "lightning",
+				Name:      "row_read_seconds",
+				Help:      "time needed to parse a row",
+				Buckets:   prometheus.ExponentialBuckets(0.001, 3.1622776601683795, 7),
+			}),
+
+		RowReadBytesHistogram: factory.NewHistogram(
+			prometheus.HistogramOpts{
+				Namespace: "lightning",
+				Name:      "row_read_bytes",
+				Help:      "number of bytes being read out from data source",
+				Buckets:   prometheus.ExponentialBuckets(1024, 2, 8),
+			}),
+
+		RowEncodeSecondsHistogram: factory.NewHistogram(
+			prometheus.HistogramOpts{
+				Namespace: "lightning",
+				Name:      "row_encode_seconds",
+				Help:      "time needed to encode a row",
+				Buckets:   prometheus.ExponentialBuckets(0.001, 3.1622776601683795, 10),
+			}),
+		RowKVDeliverSecondsHistogram: factory.NewHistogram(
+			prometheus.HistogramOpts{
+				Namespace: "lightning",
+				Name:      "row_kv_deliver_seconds",
+				Help:      "time needed to deliver kvs of a single row",
+				Buckets:   prometheus.ExponentialBuckets(0.001, 3.1622776601683795, 10),
+			}),
+
+		BlockDeliverSecondsHistogram: factory.NewHistogram(
+			prometheus.HistogramOpts{
+				Namespace: "lightning",
+				Name:      "block_deliver_seconds",
+				Help:      "time needed to deliver a block",
+				Buckets:   prometheus.ExponentialBuckets(0.001, 3.1622776601683795, 10),
+			}),
+		BlockDeliverBytesHistogram: factory.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: "lightning",
+				Name:      "block_deliver_bytes",
+				Help:      "number of bytes being sent out to importer",
+				Buckets:   prometheus.ExponentialBuckets(512, 2, 10),
+			}, []string{"kind"}),
+		BlockDeliverKVPairsHistogram: factory.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: "lightning",
+				Name:      "block_deliver_kv_pairs",
+				Help:      "number of KV pairs being sent out to importer",
+				Buckets:   prometheus.ExponentialBuckets(1, 2, 10),
+			}, []string{"kind"}),
+		ChecksumSecondsHistogram: factory.NewHistogram(
+			prometheus.HistogramOpts{
+				Namespace: "lightning",
+				Name:      "checksum_seconds",
+				Help:      "time needed to complete the checksum stage",
+				Buckets:   prometheus.ExponentialBuckets(1, 2.2679331552660544, 10),
+			}),
+
+		LocalStorageUsageBytesGauge: factory.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: "lightning",
+				Name:      "local_storage_usage_bytes",
+				Help:      "disk/memory size currently occupied by intermediate files in local backend",
+			}, []string{"medium"}),
+
+		ProgressGauge: factory.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: "lightning",
+				Name:      "progress",
+				Help:      "progress of lightning phase",
+			}, []string{"phase"}),
+	}
+}
+
+// RegisterTo registers all metrics to the given registry.
+func (m *Metrics) RegisterTo(r promutil.Registry) {
```

Comment:
maybe check nil here, so caller doesn't bother to implement it.

Reply:
check `m` or check `r`?

Reply:
check `r`, so caller can set `nil` PromRegistry  to skip register and unregister

Reply:
I think it's better to use `promutil.NewNoopRegistry()` explicitly. Otherwise, the metrics may accidentally skip `register` and `unregister`.